device/devicegate/filter.go

```diff
@@ -0,0 +1,241 @@
+package devicegate
+
+import (
+	"fmt"
+	"strings"
+	"sync"
+
+	"github.com/xmidt-org/webpa-common/device"
+)
+
+const (
+	metadataMapLocation = "metadata_map"
+	claimsLocation      = "claims"
+)
+
+// Interface is a gate interface specifically for filtering devices
+type Interface interface {
+	device.Filter
+
+	// VisitAll applies the given visitor function to each set of filter values
+	//
+	// No methods on this Interface should be called from within the visitor function, or
+	// a deadlock will likely occur.
+	VisitAll(visit func(string, Set) bool) int
+
+	// GetFilter returns the set of filter values associated with a filter key and a bool
+	// that is true if the key was found, false if it doesn't exist.
+	GetFilter(key string) (Set, bool)
+
+	// SetFilter saves the filter values and filter key to filter by. It returns a Set of the old values and a
+	// bool that is true if the filter key did not previously exist and false if the filter key had existed beforehand.
+	SetFilter(key string, values []interface{}) (Set, bool)
+
+	// DeleteFilter deletes a filter key. This completely removes all filter values associated with that key as well.
+	// Returns true if key had existed and values actually deleted, and false if key was not found.
+	DeleteFilter(key string) bool
+
+	// GetAllowedFilters returns the set of filters that devices are allowed to be filtered by. Also returns a
+	// bool that is true if there are allowed filters set, and false if there aren't (meaning that all filters are allowed)
+	GetAllowedFilters() (Set, bool)
+}
+
+// Set is an interface that represents a read-only hashset
+type Set interface {
+
+	// Has returns true if a value exists in the set, false if it doesn't.
+	Has(interface{}) bool
+
+	// VisitAll applies the visitor function to every value in the set.
+	VisitAll(func(interface{}))
+
+	// String returns a string representation of the set.
+	String() string
```

Comment:
Since the `FilterSet` is returning a json representation of the Set. Have this interface include [`json.Marshaler`](https://golang.org/pkg/encoding/json/#Marshaler)